### PR TITLE
Fix: Nested editable on Safari.

### PR DESCRIPTION
**Pull request by gorzelinski**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Fix: Nested editable on Safari. Closes https://github.com/ckeditor/ckeditor5/issues/17383.

---

### Additional information

_For example – encountered issues, assumptions you had to make, other affected tickets, etc._

### CODE REVIEW

File: packages/ckeditor5-widget/src/widget.ts [hunk 1]
```diff
@@ -251,6 +251,23 @@ export default class Widget extends Plugin {
 			}
 		}, { priority: 'low' } );
 
+		// Handle Backspace key while inside a nested editable on Safari. See #17383.
+		this.listenTo<ViewDocumentKeyDownEvent>( viewDocument, 'keydown', ( evt, data ) => {
+			if ( viewDocument.isComposing ||
+				data.keystroke != keyCodes.backspace && !data.metaKey ||
```
Comment (gorzelinski) [inline packages/ckeditor5-widget/src/widget.ts:257]: This condition solves the problem mentioned in the issue but introduces a new one. It's too broad. But I'm not sure how to make it more strict. Each time I tried, something else broke. Hmm...

**Comment by gorzelinski** [comment]
> Please add tests to cover proposed solution and to make 100% of CC.

I fixed the solution with the help of Kuba, and I am working on it :)

**Comment by gorzelinski** [comment]
The added test passes, but it's probably a false positive. I feel like I did something wrong with emulating clicking backspace, but I'm unsure what. It doesn't help that I found many ways to do that in our tests. Could you take a look in your free time? @pszczesniak

File: packages/ckeditor5-typing/src/delete.ts [hunk 2]
```diff
@@ -82,6 +84,23 @@ export default class Delete extends Plugin {
 			view.scrollToTheSelection();
 		}, { priority: 'low' } );
 
+		// Handle Backspace key while inside a nested editable on Safari. See https://github.com/ckeditor/ckeditor5/issues/17383.
```
Comment (niegowski) [inline packages/ckeditor5-typing/src/delete.ts:87]:
```suggestion
		// Handle Backspace key while at the beginning of a nested editable. See https://github.com/ckeditor/ckeditor5/issues/17383.
```

File: packages/ckeditor5-typing/src/delete.ts [hunk 3]
```diff
@@ -82,6 +84,23 @@ export default class Delete extends Plugin {
...
+		this.listenTo<ViewDocumentKeyDownEvent>( viewDocument, 'keydown', ( evt, data ) => {
+			if ( viewDocument.isComposing ||
+				data.keyCode != keyCodes.backspace ||
+				!editor.model.document.selection.isCollapsed
+			) {
```
Comment (niegowski) [inline packages/ckeditor5-typing/src/delete.ts:92]:
```suggestion
			if (
				viewDocument.isComposing ||
				data.keyCode != keyCodes.backspace ||
				!modelDocument.selection.isCollapsed
			) {
```

File: packages/ckeditor5-typing/src/delete.ts [hunk 4]
```diff
@@ -82,6 +84,23 @@ export default class Delete extends Plugin {
...
+				return;
+			}
+
+			const ancestorLimit = editor.model.schema.getLimitElement( editor.model.document.selection );
```
Comment (niegowski) [inline packages/ckeditor5-typing/src/delete.ts:96]:
```suggestion
			const ancestorLimit = editor.model.schema.getLimitElement( modelDocument.selection );
```

File: packages/ckeditor5-typing/src/delete.ts [hunk 5]
```diff
@@ -82,6 +84,23 @@ export default class Delete extends Plugin {
...
+			const limitStartPosition = editor.model.createPositionAt( ancestorLimit, 0 );
+
+			if ( limitStartPosition.isTouching( editor.model.document.selection.getFirstPosition()! ) ) {
```
Comment (niegowski) [inline packages/ckeditor5-typing/src/delete.ts:99]:
```suggestion
			if ( limitStartPosition.isTouching( modelDocument.selection.getFirstPosition()! ) ) {
```

File: packages/ckeditor5-typing/tests/delete.js [hunk 6]
```diff
@@ -76,6 +111,39 @@ describe( 'Delete feature', () => {
 		expect( spy.calledWithMatch( 'delete', { unit: 'character', sequence: 5 } ) ).to.be.true;
 	} );
 
+	// See https://github.com/ckeditor/ckeditor5/issues/17383.
+	it( 'handles the backspace key in a nested editable (Safari)', () => {
+		env.isSafari = true;
```
Comment (niegowski) [inline packages/ckeditor5-typing/tests/delete.js:116]: This is a generic solution, not aimed at Safari only.

Comment (gorzelinski) [reply packages/ckeditor5-typing/tests/delete.js:116]: I found an edge case on Safari, but you're right - the solution targets all browsers.

File: packages/ckeditor5-typing/tests/delete.js [hunk 7]
```diff
@@ -3,30 +3,67 @@
...
+
+				editor.conversion.for( 'downcast' )
+					.elementToElement( {
+						model: 'widget',
+						view: ( modelItem, { writer } ) => {
+							const b = writer.createAttributeElement( 'b' );
+							const div = writer.createContainerElement( 'div' );
+							writer.insert( writer.createPositionAt( div, 0 ), b );
+
+							return toWidget( div, writer, { label: 'element label' } );
```
Comment (niegowski) [inline packages/ckeditor5-typing/tests/delete.js:60]:
```suggestion
							const div = writer.createContainerElement( 'div' );

							return toWidget( div, writer, { label: 'element label' } );
```

File: packages/ckeditor5-typing/tests/delete.js [hunk 8]
```diff
@@ -3,30 +3,67 @@
...
+						}
+					} )
+					.elementToElement( {
+						model: 'nested',
+						view: ( modelItem, { writer } ) => writer.createEditableElement( 'figcaption', { contenteditable: true } )
```
Comment (niegowski) [inline packages/ckeditor5-typing/tests/delete.js:65]:
```suggestion
						view: ( modelItem, { writer } ) => toWidgetEditable( writer.createEditableElement( 'figcaption', { contenteditable: true } ), writer )
```

File: packages/ckeditor5-typing/tests/delete.js [hunk 9]
```diff
@@ -76,6 +113,60 @@ describe( 'Delete feature', () => {
...
+		expect( getModelData( model ) ).to.equal( '<widget><nested>[]</nested></widget>' );
+
+		expect( clickBackspace().preventedKeyDown ).to.be.true;
+
+		expect( getModelData( model ) ).to.equal( '<widget><nested>[]</nested></widget>' );
+
+		// There was an edge case tied to deleting whole lines.
+		expect( clickBackspace( true ).preventedKeyDown ).to.be.true;
+
+		expect( getModelData( model ) ).to.equal( '<widget><nested>[]</nested></widget>' );
```
Comment (niegowski) [inline packages/ckeditor5-typing/tests/delete.js:167]: I'd split those cases into separate tests.

Comment (gorzelinski) [reply packages/ckeditor5-typing/tests/delete.js]: Makes sense.

**Review by Mati365 — DISMISSED** [review]

File: packages/ckeditor5-typing/tests/delete.js (hunk 9 above)
Comment (niegowski) [reply packages/ckeditor5-typing/tests/delete.js]: I meant all of those cases, not just the last one :D

**Review by pszczesniak — DISMISSED** [review]

**Review by Mati365 — APPROVED** [review]

**Review by niegowski — APPROVED** [review]
